vitrivr-engine-index/src/main/kotlin/org/vitrivr/engine/index/decode/FFmpegVideoDecoder.kt

```diff
@@ -0,0 +1,230 @@
+                    return@collect
+                }
+
+                val probeResult = ffprobe.setShowStreams(true).also {
+                    if (source is FileSource) {
+                        it.setInput(source.path)
+                    } else {
+                        it.setInput(source.newInputStream())
+                    }
+                }.execute()
+
+                val videoStreamInfo = probeResult.streams.find { it.codecType == StreamType.VIDEO }
+
+                if (videoStreamInfo != null) {
+                    source.metadata[Metadata.METADATA_KEY_VIDEO_FPS] = videoStreamInfo.avgFrameRate.toDouble()
+                    source.metadata[Metadata.METADATA_KEY_AV_DURATION] = (videoStreamInfo.duration * 1000f).toLong()
+                    source.metadata[Metadata.METADATA_KEY_IMAGE_WIDTH] = videoStreamInfo.width
+                    source.metadata[Metadata.METADATA_KEY_IMAGE_HEIGHT] = videoStreamInfo.height
+                }
+
+                val audioStreamInfo = probeResult.streams.find { it.codecType == StreamType.AUDIO }
+
+                if (audioStreamInfo != null) {
+                    source.metadata[Metadata.METADATA_KEY_AUDIO_CHANNELS] = audioStreamInfo.channels
+                    source.metadata[Metadata.METADATA_KEY_AUDIO_SAMPLERATE] = audioStreamInfo.sampleRate
+                    source.metadata[Metadata.METADATA_KEY_AUDIO_SAMPLESIZE] = audioStreamInfo.sampleFmt
+                }
+
+                var windowEnd = TimeUnit.MILLISECONDS.toMicros(this@Instance.timeWindowMs)
+
+                val imageTransferBuffer = LinkedBlockingQueue<Pair<BufferedImage, Long>>(10)
+
+                val ffmpegInstance = ffmpeg.addInput(
+                    if (source is FileSource) {
+                        UrlInput.fromPath(source.path)
+                    } else {
+                        PipeInput.pumpFrom(source.newInputStream())
+                    }
+                ).addOutput(
+                    FrameOutput.withConsumerAlpha(
+                        object : FrameConsumer {
+
+                            val streamMap = mutableMapOf<Int, Stream>()
+
+                            override fun consumeStreams(streams: MutableList<Stream>) {
+                                streams.forEach { stream -> streamMap[stream.id] = stream }
+                            }
+
+                            override fun consume(frame: Frame) {
+
+                                val stream = streamMap[frame.streamId] ?: return
+
+                                when (stream.type) {
+                                    Stream.Type.VIDEO -> {
+                                        imageTransferBuffer.put(frame.image!! to (1000000 * frame.pts) / stream.timebase)
+                                    }
+
+                                    Stream.Type.AUDIO -> {
+                                        //TODO
```

Comment:
Is there a particular reason this is not implemented? 

I think we should try to create complete PRs in the sense that they shouldn't introduce additional TODOs nobody's ever going to take care off (unless there is a good reason to).

Reply:
I agree in principle. In this case, I barely had time to build this thing as is, but apparently, it is needed as the current decoder still appears to suffer from memory issues with longer extraction runs. Since this one spawns a new process for every file, there should at least not be any issues with accumulating memory garbage. I don't know when I will get around to implementing this, especially since it is somewhat tedious to test...

Reply:
Added this in my last commit.